src/util/finalizable_processor.rs

```diff
@@ -0,0 +1,140 @@
+        // theoratically we could do the following loop at any time in a GC (not necessarily after closure phase).
+        // But we have to iterate through candidates after closure.
+        self.candidates.append(&mut self.ready_for_finalize);
+
+        for reff in self
+            .candidates
+            .drain(start..)
+            .collect::<Vec<ObjectReference>>()
+        {
+            trace!("Pop {:?} for finalization", reff);
+            if reff.is_live() {
+                let res = FinalizableProcessor::get_forwarded_finalizable(e, reff);
+                trace!("{:?} is live, push {:?} back to candidates", reff, res);
+                self.candidates.push(res);
+                continue;
+            }
+
+            let retained = FinalizableProcessor::return_for_finalize(e, reff);
+            self.ready_for_finalize.push(retained);
+            trace!(
+                "{:?} is not live, push {:?} to ready_for_finalize",
+                reff,
+                retained
+            );
+        }
+        e.flush();
+
+        self.nursery_index = self.candidates.len();
+    }
+
+    pub fn forward<E: ProcessEdgesWork>(&mut self, e: &mut E, _nursery: bool) {
+        self.candidates
+            .iter_mut()
+            .for_each(|reff| *reff = FinalizableProcessor::get_forwarded_finalizable(e, *reff));
+        e.flush();
+    }
+
+    pub fn get_ready_object(&mut self) -> Option<ObjectReference> {
+        self.ready_for_finalize.pop()
+    }
+}
+
+#[derive(Default)]
+pub struct Finalization<E: ProcessEdgesWork>(PhantomData<E>);
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for Finalization<E> {
+    fn do_work(&mut self, worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        let mut finalizable_processor = mmtk.finalizable_processor.lock().unwrap();
+        debug!(
+            "Finalization, {} objects in candidates, {} objects ready to finalize",
+            finalizable_processor.candidates.len(),
+            finalizable_processor.ready_for_finalize.len()
+        );
+
+        let mut w = E::new(vec![], false, mmtk);
+        w.set_worker(worker);
+        finalizable_processor.scan(&mut w, mmtk.plan.in_nursery());
+        debug!(
+            "Finished finalization, {} objects in candidates, {} objects ready to finalize",
+            finalizable_processor.candidates.len(),
+            finalizable_processor.ready_for_finalize.len()
+        );
+    }
+}
+impl<E: ProcessEdgesWork> Finalization<E> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+#[derive(Default)]
+pub struct ForwardFinalization<E: ProcessEdgesWork>(PhantomData<E>);
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for ForwardFinalization<E> {
+    fn do_work(&mut self, _worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        trace!("Forward finalization");
+        let mut finalizable_processor = mmtk.finalizable_processor.lock().unwrap();
```

Comment:
If we hold this lock during the whole do_work, then this ForwardFinalization work packets will never execute in parallel if we have multiple instances of this work packet. But currently, I'm not sure whether there are any better options...

Reply:
Looks like this work packet (as well as `Finalization`) can only have one instance during a GC cycle. So it should be fine...

Reply:
`ForwardFinalization` is in the `RefForwarding` bucket, and that bucket won't open until the previous bucket `RefClosure` is done.